### PR TITLE
BUG: Ensure incomplete stata files are deleted

### PR DESCRIPTION
**Pull request by bashtage**
Attempt to delete failed writes and warn if not able to delete

- [x] closes N/A
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by bashtage** [comment]
xref #23573

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/stata.py [hunk 1]
```diff
@@ -503,6 +504,10 @@ class InvalidColumnName(Warning):
 """
 
 
+class IOWarning(Warning):
```
[the remark this answers is not in the document]
Comment (bashtage) [reply pandas/io/stata.py]: Fixed.  I wanted ResourceWarning but not available in 2.7.  Ready for 2.7 to be retired.

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/io/stata.py]: actually, use ``pandas.compat.ResourceWarning`` it is defined

**Comment by bashtage** [comment]
@jreback This is ready now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @bashtage

**Pull request merged**